domain/port/state/state.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// State represents the persistence layer for opened ports.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new state reference.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetOpenedPorts returns the opened ports for a given unit uuid,
+// grouped by endpoint.
+func (st *State) GetOpenedPorts(ctx context.Context, unit string) (network.GroupedPortRanges, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	unitUUID := unitUUID{UUID: unit}
+
+	query, err := st.Prepare(`
+SELECT &endpointPortRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+JOIN unit_endpoint ON port_range.unit_endpoint_uuid = unit_endpoint.uuid
+WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
+`, endpointPortRange{}, unitUUID)
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing get opened ports statement")
+	}
+
+	results := []endpointPortRange{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, query, unitUUID).GetAll(&results)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return domain.CoerceError(err)
+	})
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting opened ports for unit %q", unit)
+	}
+
+	groupedPortRanges := network.GroupedPortRanges{}
+	for _, endpointPortRange := range results {
+		if _, ok := groupedPortRanges[endpointPortRange.Endpoint]; !ok {
+			groupedPortRanges[endpointPortRange.Endpoint] = []network.PortRange{}
+		}
+		groupedPortRanges[endpointPortRange.Endpoint] = append(groupedPortRanges[endpointPortRange.Endpoint], decodeEndpointPortRange(endpointPortRange))
+	}
+	return groupedPortRanges, nil
+}
+
+// UpdateUnitPorts opens and closes ports for the endpoints of a given unit.
+//
+// NOTE: This method opens ports first, and then closes ports. However, it is recommended
+// that your service layer reconciles overlapping port ranges before calling this method.
+func (st *State) UpdateUnitPorts(ctx context.Context, unit string, openPorts, closePorts network.GroupedPortRanges) error {
```

Comment:
This method is quite long. Given our caching of prepared statements, there's no issue with breaking some of the steps out into separate methods, which will also bring the SQL/DML proximate to execution.

Reply:
Breaking up